src/core/sock/sockinfo_tcp.cpp

```diff
@@ -6042,3 +6024,211 @@ inline bool sockinfo_tcp::handle_bind_no_port(int &bind_ret, in_port_t in_port,
 
     return CONTINUE_WITH_BIND;
 }
+
+int sockinfo_tcp::tcp_tx_express(const struct iovec *iov, unsigned iov_len, uint32_t mkey,
+                                 xlio_express_flags flags, void *opaque_op)
+{
+    if (unlikely(!is_rts())) {
+        if (m_conn_state == TCP_CONN_TIMEOUT) {
+            si_tcp_logdbg("TX timed out");
+            errno = ETIMEDOUT;
+        } else if (m_conn_state == TCP_CONN_RESETED) {
+            si_tcp_logdbg("TX on reseted socket");
+            errno = ECONNRESET;
+        } else if (m_conn_state == TCP_CONN_ERROR) {
+            si_tcp_logdbg("TX on connection failed socket");
+            errno = ECONNREFUSED;
+        } else {
+            si_tcp_logdbg("TX on disconnected socket");
+            errno = EPIPE;
+        }
+        return -1;
+    }
+
+    err_t err;
+    pbuf_desc mdesc;
+
+    if (unlikely(!is_rts())) {
+        if (m_conn_state == TCP_CONN_TIMEOUT) {
+            si_tcp_logdbg("TX timed out");
+            errno = ETIMEDOUT;
+        } else if (m_conn_state == TCP_CONN_RESETED) {
+            si_tcp_logdbg("TX on reseted socket");
+            errno = ECONNRESET;
+        } else if (m_conn_state == TCP_CONN_ERROR) {
+            si_tcp_logdbg("TX on connection failed socket");
+            errno = ECONNREFUSED;
+        } else {
+            si_tcp_logdbg("TX on disconnected socket");
+            errno = EPIPE;
+        }
+        return -1;
+    }
+
+    switch (flags & XLIO_EXPRESS_OP_TYPE_MASK) {
+    case XLIO_EXPRESS_OP_TYPE_DESC:
+        mdesc.attr = PBUF_DESC_EXPRESS;
+        break;
+    case XLIO_EXPRESS_OP_TYPE_FILE_ZEROCOPY:
+        mdesc.attr = PBUF_DESC_MDESC;
+        break;
+    default:
+        return -1;
+    };
+    mdesc.mkey = mkey;
+    mdesc.opaque = opaque_op;
+
+    int bytes_written = 0;
+
+    lock_tcp_con();
+
+    for (unsigned i = 0; i < iov_len; ++i) {
+        err = tcp_write_express(&m_pcb, iov[i].iov_base, iov[i].iov_len, &mdesc);
+        if (err != ERR_OK) {
+            /* The only error in tcp_write_express is a memory error */
+            return tcp_tx_handle_errno_and_unlock(ENOMEM);
+        }
+        bytes_written += iov[i].iov_len;
+    }
+
+    if (!(flags & XLIO_EXPRESS_MSG_MORE)) {
+        err = tcp_output(&m_pcb);
+        if (err != ERR_OK) {
+            /* The error very likely to be recoverable */
```

Comment:
[Note for future] Need to handle error and revert all added tcp segments - we don't expect memory allocation errors, so this can be slow.

Reply:
not in context of current PR.

Reply:
I guess we need to check here if the error is ERR_MEM and then to revert posted segments. Because we must not allow partial write and in case of out of mem we have no choice.
I think it can be as separate PR. But we must handle this case somehow.

Reply:
It is simpler to set the socket/connection state to fail/error.

Reply:
This comment is misplaced - it's actually related to the tcp_write_express() call above. We must guarantee that either full iov is accepted or not accepted at all. So, we need to do that around tcp_write_express(). Once all the iov is within the stack, tcp_output() may fail and progress will be triggered later by either further send(), RX or RTO.

I see 2 possible solutions for this:
1. push iov to tcp_write_express() instead of single pointers. In such a way tcp_write_express() will fail on memory error and no revert will be needed.
2. in corner case with memory error, revert partial data as a slow path.

I'd prefer (1) if it doesn't make impact on performance. Don't do this in context of current PR. Probably it will be done in context of aes_xts

Reply:
In the current implementation, we first submit the entire iov to LwIP, and then we attempt tcp_output.
Moving the for loop (where we submit all the segments one-by-one) from sockinfo_tcp.cpp to tcp_out.c is not a big deal, as another call site already relies on a single segment press write API.

Reply:
tcp_wirte-like functions prepare all the TCP segments and only in case of no errors do actual insert into the stream and update the PCB. So, moving the iov loop will resolve the issue with partial write, since tcp_write() already has the revert mechanism in case of errors. But again, this is in plans for other PR.